File: fc-chat/demo/part3/ch3/02-5-message-system-add-connection/src/main/java/net/prostars/messagesystem/json/JsonUtil.java
package net.prostars.messagesystem.json;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class JsonUtil {

  private static final Logger log = LoggerFactory.getLogger(JsonUtil.class);

  private final ObjectMapper objectMapper;

  public JsonUtil(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  public <T> Optional<T> fromJson(String json, Class<T> clazz) {
    try {
      return Optional.of(objectMapper.readValue(json, clazz));
    } catch (Exception ex) {
      log.error("Failed JSON to Object: {}", ex.getMessage());
      return Optional.empty();
    }
  }

  public Optional<String> toJson(Object object) {
    try {
      return Optional.of(objectMapper.writeValueAsString(object));
    } catch (Exception ex) {
      log.error("Failed Object to JSON: {}", ex.getMessage());
      return Optional.empty();
    }
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/json/JsonUtil.kt
package com.yscorp.withpush.messagesystem.json

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.*

@Component
class JsonUtil(private val objectMapper: ObjectMapper) {

    /**
     * JSON 문자열을 지정된 클래스 타입의 객체로 역직렬화합니다.
     * 역직렬화에 실패하면 null을 반환합니다.
     */
    fun <T> fromJson(json: String?, clazz: Class<T>): T? {
        return runCatching {
            objectMapper.readValue(json, clazz)
        }.getOrElse {
            log.error("Failed JSON to Object: {}", it.message)
            null
        }
    }

    /**
     * 객체를 JSON 문자열로 직렬화합니다.
     * 직렬화에 실패하면 null을 반환합니다.
     */
    fun toJson(obj: Any?): String? {
        return runCatching {
            objectMapper.writeValueAsString(obj)
        }.getOrElse {
            log.error("Failed Object to JSON: {}", it.message)
            null
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(JsonUtil::class.java)
    }
}
